rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java

```diff
@@ -0,0 +1,86 @@
+package co.rsk.net.sync;
+
+import co.rsk.net.NodeID;
+import co.rsk.net.messages.BodyResponseMessage;
+import com.google.common.annotations.VisibleForTesting;
+import org.ethereum.core.Block;
+import org.ethereum.core.BlockHeader;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Queue;
+
+public class DownloadingBodiesSyncState  extends BaseSyncState {
+    private Queue<BlockHeader> pendingHeaders;
+    private Map<Long, PendingBodyResponse> pendingBodyResponses = new HashMap<>();
+
+    public DownloadingBodiesSyncState(SyncConfiguration syncConfiguration, SyncEventsHandler syncEventsHandler, SyncInformation syncInformation, Queue<BlockHeader> pendingHeaders) {
+        super(syncInformation, syncEventsHandler, syncConfiguration);
+
+        this.pendingHeaders = pendingHeaders;
+    }
+
+    @Override
+    public void newBody(BodyResponseMessage message) {
+        if (!isExpectedBody(message.getId())) {
+            // Invalid body response
+            // TODO(mc) do peer scoring, banning
+            syncEventsHandler.onErrorSyncing(
+                    "Unexpected body received from node {}",
+                    syncInformation.getSelectedPeerId());
+            return;
+        }
+
+        // we know it exists because it was called from a SyncEvent
+        BlockHeader header = pendingBodyResponses.get(message.getId()).header;
+        Block block = Block.fromValidData(header, message.getTransactions(), message.getUncles());
+        syncInformation.processBlock(block);
+
+        if (!pendingHeaders.isEmpty()) {
+            requestBody();
+            return;
+        }
+
+        // Finished syncing
+        syncEventsHandler.onCompletedSyncing();
+    }
+
+    @Override
+    public void onEnter() {
+        requestBody();
+    }
+
+    private void requestBody(){
+        BlockHeader header = pendingHeaders.remove();
```

Comment:
And if the respective request fails... ?

Reply:
Can be improved but if the request fails the process should restart.